tensorflow-training/src/main/java/org/tensorflow/training/optimizers/Optimizer.java

```diff
@@ -0,0 +1,279 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.tensorflow.training.optimizers;
+
+import org.tensorflow.Graph;
+import org.tensorflow.Operand;
+import org.tensorflow.Operation;
+import org.tensorflow.Output;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.Scope;
+import org.tensorflow.op.core.Assign;
+import org.tensorflow.op.core.NoOp;
+import org.tensorflow.op.core.Variable;
+import org.tensorflow.types.family.TType;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ *
+ */
+public abstract class Optimizer {
+
+  public static final String VARIABLE_V2 = "VariableV2";
+
+  /**
+   * Optional attributes for {@link org.tensorflow.training.optimizers.Optimizer}
+   */
+  public static class Options {
+
+    /**
+     * @param sharedName If non-empty, this variable is named in the given bucket
+     * with this shared_name. Otherwise, the node name is used instead.
+     */
+    public Optimizer.Options sharedName(String sharedName) {
+      this.sharedName = sharedName;
+      return this;
+    }
+
+    protected String sharedName;
+
+    private Options() {
+    }
+  }
+  /**
+   * Top level map key is the variable name, lower level map key is the slot name.
+   */
+  private final Map<String, Map<String, Variable<?>>> slots;
+
+  /**
+   * Global state variables
+   */
+  //TODO make this be used.
+  protected final List<Variable<?>> globals;
+
+  /**
+   * The Graph this optimizer is operating on.
+   */
+  protected final Graph graph;
+
+  /**
+   * The ops builder for the graph.
+   */
+  protected final Ops tf;
+
+  protected Optimizer(Graph graph) {
+    this.graph = graph;
+    this.tf = Ops.create(graph).withName(getOptimizerName());
+    this.slots = new HashMap<>();
+    this.globals = new ArrayList<>();
+  }
+
+  public Op minimize(Operand<?> loss) {
+    return minimize(loss, getOptimizerName() + "-minimize");
+  }
+
+  public Op minimize(Operand<?> loss, String name) {
+    List<GradAndVar<?>> gradsAndVars = computeGradients(loss);
+
+    return applyGradients(gradsAndVars, name);
+  }
+
+  public <T extends TType> List<GradAndVar<?>> computeGradients(Operand<?> loss) {
+    List<Operation> variables = new ArrayList<>();
+    Iterator<Operation> opItr = graph.operations();
+    while (opItr.hasNext()) {
+      Operation op = opItr.next();
```

Comment:
`graph.operations().forEachRemaining` maybe?

Also, is it OK that an optimizer is always applied to all variables in the graph? Is this true for all kind of graphs?

Reply:
That's how they work in Python. It's really hard to make it work any other way without a lot of ceremony.